core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -130,7 +147,64 @@ public String toString() {
      */
     @Override
     public String get(String path) {
-        return "IMDS";
+
+        String data = null;
+        AbortableInputStream abortableInputStream = null;
+        try {
+            String token = getToken();
+            URI uri = URI.create(endpoint + path);
+            HttpExecuteRequest httpExecuteRequest = requestMarshaller.createDataRequest(uri , SdkHttpMethod.GET, token,
+                                                                                tokenTtl);
+            HttpExecuteResponse response =  httpClient.prepareRequest(httpExecuteRequest).call();
+            int statusCode = response.httpResponse().statusCode();
+
+            if (statusCode == HttpURLConnection.HTTP_OK) {
+                abortableInputStream = response.responseBody().get();
+                data = IoUtils.toUtf8String(abortableInputStream);
+            } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
+                throw SdkServiceException.builder().message("The requested metadata is not found ").build();
+            } else {
+                throw SdkClientException.builder()
+                                        .message("Unexpected Exception occurred with Status Code " + statusCode).build();
+            }
+        } catch (SdkServiceException sd) {
+            throw SdkServiceException.builder().message(sd.getMessage()).cause(sd).build();
+        } catch (IOException | SdkClientException  io) {
+            // TODO Retry Logic will be added
+            log.warn("Received an IOException " + io);
+        } finally {
+            IoUtils.closeQuietly(abortableInputStream, log);
+        }
+
+        return data;
+    }
+
+    private String getToken() throws IOException {
+
+        AbortableInputStream abortableInputStream = null;
+        try {
+            URI uri = URI.create(endpoint + TOKEN_RESOURCE_PATH);
+            HttpExecuteRequest httpExecuteRequest = requestMarshaller.createTokenRequest(uri , SdkHttpMethod.PUT, tokenTtl);
+            HttpExecuteResponse response =  httpClient.prepareRequest(httpExecuteRequest).call();
+            int statusCode = response.httpResponse().statusCode();
+
+            if (statusCode == HttpURLConnection.HTTP_OK) {
+                abortableInputStream = response.responseBody().get();
+                String token = IoUtils.toUtf8String(abortableInputStream);
+                return token;
+            } else if (statusCode == HttpURLConnection.HTTP_FORBIDDEN || statusCode == HttpURLConnection.HTTP_BAD_REQUEST) {
+                throw SdkServiceException.builder().message("Could not retrieve token").build();
```

Comment:
Should we log the status code? Any reason to have this be a different case than just the else clause below?

Reply:
The status codes mentioned , should not have retries enabled on them. But for the below else case , retries can happen.